File: zhukogpt/mathfmt.py
"""Превращает LaTeX-формулы из ответов моделей ($17 \\times 3$) в обычный текст (17 × 3)."""
import re

_SYMBOLS = {
    r"\times": "×", r"\cdot": "·", r"\div": "÷", r"\pm": "±", r"\mp": "∓",
    r"\le": "≤", r"\leq": "≤", r"\ge": "≥", r"\geq": "≥", r"\ne": "≠", r"\neq": "≠",
    r"\approx": "≈", r"\equiv": "≡", r"\infty": "∞", r"\to": "→", r"\rightarrow": "→",
    r"\Rightarrow": "⇒", r"\leftarrow": "←", r"\Leftrightarrow": "⇔", r"\degree": "°",
    r"\circ": "°", r"\angle": "∠", r"\perp": "⊥", r"\parallel": "∥", r"\triangle": "△",
    r"\in": "∈", r"\notin": "∉", r"\cup": "∪", r"\cap": "∩", r"\subset": "⊂",
    r"\emptyset": "∅", r"\forall": "∀", r"\exists": "∃", r"\neg": "¬", r"\land": "∧",
    r"\lor": "∨", r"\sum": "∑", r"\prod": "∏", r"\int": "∫", r"\partial": "∂",
    r"\alpha": "α", r"\beta": "β", r"\gamma": "γ", r"\delta": "δ", r"\Delta": "Δ",
    r"\epsilon": "ε", r"\varepsilon": "ε", r"\theta": "θ", r"\lambda": "λ", r"\mu": "μ",
    r"\pi": "π", r"\rho": "ρ", r"\sigma": "σ", r"\Sigma": "Σ", r"\tau": "τ", r"\phi": "φ",
    r"\varphi": "φ", r"\omega": "ω", r"\Omega": "Ω", r"\%": "%", r"\,": " ", r"\;": " ",
    r"\:": " ", r"\!": "", r"\quad": "  ", r"\qquad": "    ", r"\ldots": "…", r"\dots": "…",
    r"\cdots": "⋯", r"\{": "{", r"\}": "}", r"\ ": " ", r"\$": "$",
}
_SUPERSCRIPT = str.maketrans("0123456789+-=()n", "⁰¹²³⁴⁵⁶⁷⁸⁹⁺⁻⁼⁽⁾ⁿ")
_SUBSCRIPT = str.maketrans("0123456789+-=()", "₀₁₂₃₄₅₆₇₈₉₊₋₌₍₎")

# Внутри скобок одного уровня — для \frac{..}{..}, \sqrt{..} и т. п.
_ARG = r"\{((?:[^{}]|\{[^{}]*\})*)\}"


def _script(text, table):
    converted = text.translate(table)
    # Если хотя бы один символ не нашёлся в таблице — оставляем через ^( ) / _( ), чтобы не потерять смысл.
    return converted if all(c != o or c in " " for c, o in zip(converted, text)) else None


def _latex_to_text(expr: str) -> str:
    s = expr
    s = re.sub(r"\^\s*\{?\s*\\circ\s*\}?", "°", s)  # 90^\circ → 90°
    for _ in range(3):  # вложенные \frac{\sqrt{..}}{..}: корни раньше дробей
        s = re.sub(r"\\sqrt\[([^\]]+)\]" + _ARG, lambda m: f"{m[1]}√({m[2]})", s)
        s = re.sub(r"\\sqrt" + _ARG, lambda m: f"√{_wrap(m[1])}", s)
        s = re.sub(r"\\[dt]?frac" + _ARG + _ARG, lambda m: f"{_wrap(m[1])}/{_wrap(m[2])}", s)
        s = re.sub(r"\\(?:text|textbf|textit|mathrm|mathbf|operatorname|boxed|mbox)" + _ARG, r"\1", s)
        s = re.sub(r"\\(?:overline|bar)" + _ARG, r"\1̄", s)
        s = re.sub(r"\\vec" + _ARG, r"\1⃗", s)
    s = re.sub(r"\\(?:left|right|big|Big|bigg|Bigg)\s*([()\[\]|.]|\\[{}])",
               lambda m: "" if m[1] == "." else m[1].lstrip("\\"), s)
    s = re.sub(r"\\(?:sin|cos|tan|cot|log|ln|lg|lim|max|min|exp)\b", lambda m: m[0][1:], s)
    for cmd in sorted(_SYMBOLS, key=len, reverse=True):  # длинные раньше: \leq до \le
        s = re.sub(re.escape(cmd) + r"(?![A-Za-z])", lambda _m, v=_SYMBOLS[cmd]: v, s)

    def sup(m):
        body = m[1] if m[1] is not None else m[2]
        return _script(body, _SUPERSCRIPT) or (f"^{body}" if len(body) == 1 else f"^({body})")

    def sub(m):
        body = m[1] if m[1] is not None else m[2]
        return _script(body, _SUBSCRIPT) or (f"_{body}" if len(body) == 1 else f"_({body})")

    s = re.sub(r"\^\{([^{}]*)\}|\^([0-9A-Za-z+\-])", sup, s)
    s = re.sub(r"_\{([^{}]*)\}|_([0-9])", sub, s)
    s = s.replace("{", "").replace("}", "")
    return re.sub(r"[ \t]{2,}", " ", s).strip()


def _wrap(part: str) -> str:
    """Скобки вокруг сложного числителя/знаменателя: (a+b)/2, но 1/2."""
    part = part.strip()
    return part if re.fullmatch(r"[\w.,√²³]+", part) else f"({part})"


def latex_to_plain(text: str) -> str:
    """Находит формулы ($…$, $$…$$, \\(…\\), \\[…\\]) и переводит их в обычный текст."""
    if "$" not in text and "\\" not in text:
        return text
    text = re.sub(r"\$\$(.+?)\$\$", lambda m: _latex_to_text(m[1]), text, flags=re.S)
    text = re.sub(r"\\\[(.+?)\\\]", lambda m: _latex_to_text(m[1]), text, flags=re.S)
    text = re.sub(r"\\\((.+?)\\\)", lambda m: _latex_to_text(m[1]), text, flags=re.S)
    # $…$ в одной строке; «$5 и $10» (цены) не трогаем: после открывающего $ нет пробела,
    # перед закрывающим нет пробела, а после него не цифра.
    text = re.sub(r"(?<![\\$])\$(?=\S)([^$\n]+?)(?<=\S)\$(?!\d)", lambda m: _latex_to_text(m[1]), text)
    return text


def keep_line_breaks(text: str) -> str:
    """В Markdown одиночный перенос склеивает строки — делаем его настоящим переносом («  » в конце).

    Иначе «76 + 1 = 77» и «74 + 1 = 75» на соседних строках сливаются в одну. Код в ``` не трогаем.
    """
    lines = text.split("\n")
    in_code = False
    for i, line in enumerate(lines[:-1]):
        if line.lstrip().startswith("```"):
            in_code = not in_code
            continue
        if not in_code and line.strip() and lines[i + 1].strip():
            lines[i] = line.rstrip() + "  "
    return "\n".join(lines)

File: web/mathfmt.js
// Порт zhukogpt/mathfmt.py: LaTeX-формулы из ответов ($17 \times 3$) → обычный текст (17 × 3),
// плюс маленький безопасный Markdown-рендер для ответа (HTML из ответа модели всегда экранируется).

const SYMBOLS = {
  "\\times": "×", "\\cdot": "·", "\\div": "÷", "\\pm": "±", "\\mp": "∓",
  "\\le": "≤", "\\leq": "≤", "\\ge": "≥", "\\geq": "≥", "\\ne": "≠", "\\neq": "≠",
  "\\approx": "≈", "\\equiv": "≡", "\\infty": "∞", "\\to": "→", "\\rightarrow": "→",
  "\\Rightarrow": "⇒", "\\leftarrow": "←", "\\Leftrightarrow": "⇔", "\\degree": "°",
  "\\circ": "°", "\\angle": "∠", "\\perp": "⊥", "\\parallel": "∥", "\\triangle": "△",
  "\\in": "∈", "\\notin": "∉", "\\cup": "∪", "\\cap": "∩", "\\subset": "⊂",
  "\\emptyset": "∅", "\\forall": "∀", "\\exists": "∃", "\\neg": "¬", "\\land": "∧",
  "\\lor": "∨", "\\sum": "∑", "\\prod": "∏", "\\int": "∫", "\\partial": "∂",
  "\\alpha": "α", "\\beta": "β", "\\gamma": "γ", "\\delta": "δ", "\\Delta": "Δ",
  "\\epsilon": "ε", "\\varepsilon": "ε", "\\theta": "θ", "\\lambda": "λ", "\\mu": "μ",
  "\\pi": "π", "\\rho": "ρ", "\\sigma": "σ", "\\Sigma": "Σ", "\\tau": "τ", "\\phi": "φ",
  "\\varphi": "φ", "\\omega": "ω", "\\Omega": "Ω", "\\%": "%", "\\,": " ", "\\;": " ",
  "\\:": " ", "\\!": "", "\\quad": "  ", "\\qquad": "    ", "\\ldots": "…", "\\dots": "…",
  "\\cdots": "⋯", "\\{": "{", "\\}": "}", "\\ ": " ", "\\$": "$",
};
const SUP = { "0": "⁰", "1": "¹", "2": "²", "3": "³", "4": "⁴", "5": "⁵", "6": "⁶", "7": "⁷", "8": "⁸",
  "9": "⁹", "+": "⁺", "-": "⁻", "=": "⁼", "(": "⁽", ")": "⁾", "n": "ⁿ" };
const SUB = { "0": "₀", "1": "₁", "2": "₂", "3": "₃", "4": "₄", "5": "₅", "6": "₆", "7": "₇", "8": "₈",
  "9": "₉", "+": "₊", "-": "₋", "=": "₌", "(": "₍", ")": "₎" };
const ARG = "\\{((?:[^{}]|\\{[^{}]*\\})*)\\}";
const SYMBOL_KEYS = Object.keys(SYMBOLS).sort((a, b) => b.length - a.length);
const escapeRe = (s) => s.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");

function script(text, table) {
  let out = "";
  for (const ch of text) {
    if (ch === " ") { out += ch; continue; }
    if (!(ch in table)) return null;  // не нашли символ — оставим ^( ) / _( ), чтобы не потерять смысл
    out += table[ch];
  }
  return out;
}

function wrap(part) {
  part = part.trim();
  return /^[\p{L}\p{N}_.,√²³]+$/u.test(part) ? part : `(${part})`;
}

function latexToText(expr) {
  let s = expr.replace(/\^\s*\{?\s*\\circ\s*\}?/g, "°");
  for (let i = 0; i < 3; i++) {  // вложенные \frac{\sqrt{..}}{..}: корни раньше дробей
    s = s.replace(new RegExp("\\\\sqrt\\[([^\\]]+)\\]" + ARG, "g"), (_, n, b) => `${n}√(${b})`);
    s = s.replace(new RegExp("\\\\sqrt" + ARG, "g"), (_, b) => `√${wrap(b)}`);
    s = s.replace(new RegExp("\\\\[dt]?frac" + ARG + ARG, "g"), (_, a, b) => `${wrap(a)}/${wrap(b)}`);
    s = s.replace(new RegExp("\\\\(?:text|textbf|textit|mathrm|mathbf|operatorname|boxed|mbox)" + ARG, "g"), "$1");
    s = s.replace(new RegExp("\\\\(?:overline|bar)" + ARG, "g"), "$1̄");
    s = s.replace(new RegExp("\\\\vec" + ARG, "g"), "$1⃗");
  }
  s = s.replace(/\\(?:left|right|big|Big|bigg|Bigg)\s*([()[\]|.]|\\[{}])/g,
    (_, b) => (b === "." ? "" : b.replace(/^\\/, "")));
  s = s.replace(/\\(sin|cos|tan|cot|log|ln|lg|lim|max|min|exp)\b/g, "$1");
  for (const cmd of SYMBOL_KEYS) {
    s = s.replace(new RegExp(escapeRe(cmd) + "(?![A-Za-z])", "g"), () => SYMBOLS[cmd]);
  }
  s = s.replace(/\^\{([^{}]*)\}|\^([0-9A-Za-z+\-])/g, (_, a, b) => {
    const body = a ?? b;
    return script(body, SUP) ?? (body.length === 1 ? `^${body}` : `^(${body})`);
  });
  s = s.replace(/_\{([^{}]*)\}|_([0-9])/g, (_, a, b) => {
    const body = a ?? b;
    return script(body, SUB) ?? (body.length === 1 ? `_${body}` : `_(${body})`);
  });
  return s.replace(/[{}]/g, "").replace(/[ \t]{2,}/g, " ").trim();
}

export function latexToPlain(text) {
  if (!text.includes("$") && !text.includes("\\")) return text;
  text = text.replace(/\$\$([\s\S]+?)\$\$/g, (_, e) => latexToText(e));
  text = text.replace(/\\\[([\s\S]+?)\\\]/g, (_, e) => latexToText(e));
  text = text.replace(/\\\(([\s\S]+?)\\\)/g, (_, e) => latexToText(e));
  // $…$ в одной строке; «$5 и $10» (цены) не трогаем.
  text = text.replace(/(?<![\\$])\$(?=\S)([^$\n]+?)(?<=\S)\$(?!\d)/g, (_, e) => latexToText(e));
  return text;
}

const escapeHtml = (s) => s.replace(/[&<>"']/g, (c) =>
  ({ "&": "&amp;", "<": "&lt;", ">": "&gt;", '"': "&quot;", "'": "&#39;" })[c]);

function inline(text) {
  // Код сначала вынимаем, чтобы звёздочки внутри него не стали жирным.
  const codes = [];
  let s = escapeHtml(text).replace(/`([^`]+)`/g, (_, c) => { codes.push(c); return `\u0000${codes.length - 1}\u0000`; });
  s = s.replace(/\*\*(.+?)\*\*/g, "<b>$1</b>").replace(/__(.+?)__/g, "<b>$1</b>");
  s = s.replace(/(^|[^*\w])\*(?!\s)(.+?)(?<!\s)\*(?!\w)/g, "$1<i>$2</i>");
  s = s.replace(/\u0000(\d+)\u0000/g, (_, i) => `<code>${codes[i]}</code>`);
  return s;
}

/** Markdown ответа → HTML: заголовки, списки, цитаты, код, жирный/курсив; каждая строка — своя строка. */
export function renderMarkdown(text) {
  const out = [];
  let list = null;  // "ul" | "ol"
  let code = null;
  const closeList = () => { if (list) { out.push(`</${list}>`); list = null; } };
  for (const line of latexToPlain(text).replace(/\r/g, "").split("\n")) {
    if (code !== null) {
      if (line.trim().startsWith("```")) { out.push(`<pre><code>${escapeHtml(code.join("\n"))}</code></pre>`); code = null; }
      else code.push(line);
      continue;
    }
    if (line.trim().startsWith("```")) { closeList(); code = []; continue; }
    let m;
    if ((m = line.match(/^\s*[-*•]\s+(.*)$/))) {
      if (list !== "ul") { closeList(); out.push("<ul>"); list = "ul"; }
      out.push(`<li>${inline(m[1])}</li>`);
    } else if ((m = line.match(/^\s*(\d+)[.)]\s+(.*)$/))) {
      if (list !== "ol") { closeList(); out.push(`<ol start="${m[1]}">`); list = "ol"; }
      out.push(`<li>${inline(m[2])}</li>`);
    } else if ((m = line.match(/^\s*#{1,6}\s+(.*)$/))) {
      closeList(); out.push(`<h3>${inline(m[1])}</h3>`);
    } else if ((m = line.match(/^\s*>\s?(.*)$/))) {
      closeList(); out.push(`<blockquote>${inline(m[1])}</blockquote>`);
    } else if (!line.trim()) {
      closeList(); out.push('<div class="gap"></div>');
    } else {
      closeList(); out.push(`<p>${inline(line)}</p>`);
    }
  }
  if (code !== null) out.push(`<pre><code>${escapeHtml(code.join("\n"))}</code></pre>`);
  closeList();
  return out.join("");
}
